crates/iceberg/src/expr/visitors/expression_evaluator.rs

```diff
@@ -0,0 +1,819 @@
+    }
+}
+
+/// Acts as a visitor for [`ExpressionEvaluator`] to apply
+/// evaluation logic to different parts of a data structure,
+/// specifically for data file partitions.
+#[derive(Debug)]
+struct ExpressionEvaluatorVisitor<'a> {
+    /// Reference to an [`ExpressionEvaluator`].
+    expression_evaluator: &'a ExpressionEvaluator,
+    /// Reference to a [`DataFile`]'s partition [`Struct`].
+    partition: &'a Struct,
+}
+
+impl<'a> ExpressionEvaluatorVisitor<'a> {
+    /// Creates a new [`ExpressionEvaluatorVisitor`].
+    fn new(expression_evaluator: &'a ExpressionEvaluator, partition: &'a Struct) -> Self {
+        Self {
+            expression_evaluator,
+            partition,
+        }
+    }
+
+    /// Checks if the [`PrimitiveLiteral`] is null.
+    fn is_null(literal: &PrimitiveLiteral) -> bool {
+        if let PrimitiveLiteral::Boolean(false) = literal {
+            return true;
+        };
+
+        false
+    }
+}
+
+impl BoundPredicateVisitor for ExpressionEvaluatorVisitor<'_> {
+    type T = bool;
+
+    fn always_true(&mut self) -> Result<bool> {
+        Ok(true)
+    }
+
+    fn always_false(&mut self) -> Result<bool> {
+        Ok(false)
+    }
+
+    fn and(&mut self, lhs: bool, rhs: bool) -> Result<bool> {
+        Ok(lhs && rhs)
+    }
+
+    fn or(&mut self, lhs: bool, rhs: bool) -> Result<bool> {
+        Ok(lhs || rhs)
+    }
+
+    fn not(&mut self, _inner: bool) -> Result<bool> {
+        Err(Error::new(ErrorKind::Unexpected, "The evaluation of expressions should not be performed against Predicates that contain a Not operator. Ensure that \"Rewrite Not\" gets applied to the originating Predicate before binding it."))
+    }
+
+    fn is_null(&mut self, reference: &BoundReference, _predicate: &BoundPredicate) -> Result<bool> {
+        let datum = reference.accessor().get(self.partition)?;
+
+        Ok(Self::is_null(datum.literal()))
+    }
+
+    fn not_null(
+        &mut self,
+        reference: &BoundReference,
+        _predicate: &BoundPredicate,
+    ) -> Result<bool> {
+        let datum = reference.accessor().get(self.partition)?;
+
+        Ok(!Self::is_null(datum.literal()))
+    }
+
+    fn is_nan(&mut self, reference: &BoundReference, _predicate: &BoundPredicate) -> Result<bool> {
+        let datum = reference.accessor().get(self.partition)?;
+
+        Ok(datum.is_nan())
+    }
+
+    fn not_nan(&mut self, reference: &BoundReference, _predicate: &BoundPredicate) -> Result<bool> {
+        let datum = reference.accessor().get(self.partition)?;
+
+        Ok(!datum.is_nan())
+    }
+
+    fn less_than(
+        &mut self,
+        reference: &BoundReference,
+        literal: &Datum,
+        _predicate: &BoundPredicate,
+    ) -> Result<bool> {
+        let datum = reference.accessor().get(self.partition)?;
+
+        if Self::is_null(datum.literal()) {
+            return Ok(false);
+        }
+
+        Ok(datum.literal() < literal.literal())
```

Comment:
I'm not sure about removing `Ord` from `PrimitiveLiteral` since its also required by `Literal`. However, I think I get your point, so impl `PartialOrd` for `Datum` and then comparing those intstead of the primitive literals makes sense.

Reply:
I think it's best to make the changes as outlined in #378 to unblock this PR.

Reply:
@ZENOTME thanks again for the fix in #378
@liurenjie1024 I updated this PR to compare `Datum` instead of `PrimitiveLiteral`.
Once #379 is done I can make the final change and we can hopefully merge this.